river/cluster/dbstream.py

```diff
@@ -0,0 +1,416 @@
+    ----------
+    [^1]: Michael Hahsler and Matthew Bolanos (2016, pp 1449-1461). Clsutering Data Streams Based on
+          Shared Density between Micro-Clusters, IEEE Transactions on Knowledge and Data Engineering 28(6) .
+          In Proceedings of the Sixth SIAM International Conference on Data Mining,
+          April 20–22, 2006, Bethesda, MD, USA.
+    [^2]: Ester et al (1996). A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases
+          with Noise. In KDD-96 Proceedings, AAAI.
+
+    Examples
+    ----------
+
+    >>> from river import cluster
+    >>> from river import stream
+
+    >>> X = [
+    ...     [1, 0.5], [1, 0.625], [1, 0.75], [1, 1.125], [1, 1.5], [1, 1.75],
+    ...     [4, 1.5], [4, 2.25], [4, 2.5], [4, 3], [4, 3.25], [4, 3.5]
+    ... ]
+
+    >>> dbstream = cluster.DBSTREAM(clustering_threshold = 1.5,
+    ...                             fading_factor = 0.05,
+    ...                             cleanup_interval = 4,
+    ...                             intersection_factor = 0.5,
+    ...                             minimum_weight = 1)
+
+    >>> for x, _ in stream.iter_array(X):
+    ...     dbstream = dbstream.learn_one(x)
+
+    >>> dbstream.predict_one({0: 1, 1: 2})
+    0
+
+    >>> dbstream.predict_one({0: 5, 1: 2})
+    1
+
+    >>> dbstream.n_clusters
+    2
+    """
+
+    def __init__(
+        self,
+        clustering_threshold: float = 1,
+        fading_factor: float = 0.01,
+        cleanup_interval: float = 2,
+        intersection_factor: float = 0.3,
+        minimum_weight: float = 1,
+    ):
+        super().__init__()
+        self.time_stamp = 0
+
+        self.clustering_threshold = clustering_threshold
+        self.fading_factor = fading_factor
+        self.cleanup_interval = cleanup_interval
+        self.minimum_weight = minimum_weight
+        self.intersection_factor = intersection_factor
+
+        self.n_clusters = 0
+        self.clusters = {}
+        self.centers = {}
+        self.micro_clusters = {}
+
+        self.s = {}
+        self.s_t = {}
+
+    @staticmethod
+    def _distance(point_a, point_b):
+        return math.sqrt(utils.math.minkowski_distance(point_a, point_b, 2))
+
+    def _find_fixed_radius_nn(self, x):
+        fixed_radius_nn = {}
+        for i in self.micro_clusters.keys():
+            if (
+                self._distance(self.micro_clusters[i].center, x)
+                < self.clustering_threshold
+            ):
+                fixed_radius_nn[i] = self.micro_clusters[i]
+        return fixed_radius_nn
+
+    def _gaussian_neighborhood(self, point_a, point_b):
+        distance = self._distance(point_a, point_b)
+        sigma = self.clustering_threshold / 3
+        gaussian_neighborhood = math.exp(-(distance * distance) / (2 * (sigma * sigma)))
+        return gaussian_neighborhood
+
+    def _update(self, x):
+        # Algorithm 1 of Michael Hahsler and Matthew Bolanos
+
+        N = self._find_fixed_radius_nn(x)
```

Comment:
I would prefer to name it as `neighbor_clusters`, since the purpose of the function `_find_fixed_radius_nn` is to find clusters whose centers are within `self.clustering_threshold` from that point.

Reply:
agree